src/ucp/tag/tag_send.c

```diff
@@ -11,183 +11,81 @@
-            flag_iov_single) {
-            req->send.uct.func   = proto->zcopy_single;
-            UCS_PROFILE_REQUEST_EVENT(req, "start_egr_zcopy_single", req->send.length);
-        } else {
-            req->send.uct.func   = proto->zcopy_multi;
-            UCS_PROFILE_REQUEST_EVENT(req, "start_egr_zcopy_multi", req->send.length);
+            return seg_size;
         }
+        /* Fall through */
+    case UCP_DATATYPE_CONTIG: 
+        return ucs_min(rndv_rma_thresh, rndv_am_thresh);
+    case UCP_DATATYPE_GENERIC:
+        return rndv_am_thresh;
+    default:
+        ucs_error("Invalid data type %lx", req->send.datatype);
     }
-    return UCS_OK;
+ 
+    return SIZE_MAX;
 }
 
-static void ucp_tag_req_start_generic(ucp_request_t *req, size_t count,
-                                      size_t rndv_rma_thresh, size_t rndv_am_thresh,
-                                      const ucp_proto_t *proto)
+static UCS_F_ALWAYS_INLINE int
+ucp_tag_is_iov_single(ucp_datatype_t datatype, size_t count, size_t max_iov)
 {
-    ucp_ep_config_t *config = ucp_ep_config(req->send.ep);
-    ucp_dt_generic_t *dt_gen;
-    size_t length;
-    void *state;
-
-    dt_gen = ucp_dt_generic(req->send.datatype);
-    state = dt_gen->ops.start_pack(dt_gen->context, req->send.buffer, count);
-
-    req->send.state.dt.generic.state = state;
-    req->send.length = length = dt_gen->ops.packed_size(state);
-
-    if (length <= config->tag.eager.max_bcopy - proto->only_hdr_size) {
-        /* bcopy single */
-        req->send.uct.func = proto->bcopy_single;
-        UCS_PROFILE_REQUEST_EVENT(req, "start_gen_bcopy_single", req->send.length);
-    } else if (length >= rndv_am_thresh) {
-        /* rendezvous */
-        ucp_tag_send_start_rndv(req);
-        UCS_PROFILE_REQUEST_EVENT(req, "start_rndv", req->send.length);
-    } else {
-        /* bcopy multi */
-        req->send.uct.func = proto->bcopy_multi;
-        UCS_PROFILE_REQUEST_EVENT(req, "start_gen_bcopy_multi", req->send.length);
+    if (ucs_unlikely(UCP_DT_IS_IOV(datatype))) {
+        return count <= max_iov;
     }
+    return 1;
 }
 
-static inline ucs_status_ptr_t
-ucp_tag_send_req(ucp_request_t *req, size_t count, ssize_t max_short,
-                 size_t *zcopy_thresh, size_t rndv_rma_thresh, size_t rndv_am_thresh,
+static UCS_F_ALWAYS_INLINE ucs_status_ptr_t
+ucp_tag_send_req(ucp_request_t *req, size_t count,
+                 const ucp_ep_msg_config_t* msg_config,
+                 size_t rndv_rma_thresh, size_t rndv_am_thresh,
                  ucp_send_callback_t cb, const ucp_proto_t *proto)
 {
-    ucs_status_t status;
-
-    switch (req->send.datatype & UCP_DATATYPE_CLASS_MASK) {
-    case UCP_DATATYPE_CONTIG:
-    case UCP_DATATYPE_IOV:
-        status = ucp_tag_req_start(req, count, max_short, zcopy_thresh,
-                                   rndv_rma_thresh, rndv_am_thresh, proto);
+    int flag_iov_single = ucp_tag_is_iov_single(req->send.datatype, count,
```

Comment:
looks like the name 'iov_single' was wrong in the first place, since it compares to msg_config->max_iov and not 1.
also, why not put the code of this function inside ucp_tag_get_rndv_threshold()?

Reply:
👍 this variable was used in several places in original code. i have not noticed as simplified it during refactoring :)